pkg/kubelet/rkt/rkt.go

```diff
@@ -142,3 +148,248 @@ func (r *Runtime) runCommand(args ...string) ([]string, error) {
 	}
 	return strings.Split(strings.TrimSpace(string(output)), "\n"), nil
 }
+
+type resource struct {
+	limit   string
+	request string
+}
+
+// setIsolators overrides the isolators of the pod manifest if necessary.
+func setIsolators(app *appctypes.App, c *api.Container) error {
+	var isolator appctypes.Isolator
+	if len(c.Capabilities.Add) > 0 || len(c.Capabilities.Drop) > 0 || len(c.Resources.Limits) > 0 || len(c.Resources.Requests) > 0 {
+		app.Isolators = []appctypes.Isolator{}
+	}
+
+	// Retained capabilities/privileged.
+	privileged := false
+	if capabilities.Get().AllowPrivileged {
+		privileged = c.Privileged
+	} else if c.Privileged {
+		glog.Errorf("rkt: Privileged is disallowed globally")
+		// TODO(yifan): Return error?
+	}
+	var addCaps string
+	if privileged {
+		addCaps = getAllCapabilities()
+	} else {
+		addCaps = getCapabilities(c.Capabilities.Add)
+	}
+	if len(addCaps) > 0 {
+		// TODO(yifan): Replace with constructor, see:
+		// https://github.com/appc/spec/issues/268
+		value := []byte(fmt.Sprintf(`{"name":"os/linux/capabilities-retain-set","value":{"set":[%s]}`, addCaps))
+		if err := isolator.UnmarshalJSON(value); err != nil {
+			glog.Errorf("rkt: Cannot unmarshal the retained capabilites %q: %v", value, err)
+			return err
+		}
+		app.Isolators = append(app.Isolators, isolator)
+	}
+
+	// Removed capabilities.
+	dropCaps := getCapabilities(c.Capabilities.Drop)
+	if len(dropCaps) > 0 {
+		// TODO(yifan): Replace with constructor, see:
+		// https://github.com/appc/spec/issues/268
+		value := []byte(fmt.Sprintf(`{"name":"os/linux/capabilities-remove-set","value":{"set":[%s]}`, dropCaps))
+		if err := isolator.UnmarshalJSON(value); err != nil {
+			glog.Errorf("rkt: Cannot unmarshal the retained capabilites %q: %v", value, err)
+			return err
+		}
+		app.Isolators = append(app.Isolators, isolator)
+	}
+
+	// Resources.
+	resources := make(map[api.ResourceName]resource)
+	for name, quantity := range c.Resources.Limits {
+		resources[name] = resource{limit: quantity.String()}
+	}
+	for name, quantity := range c.Resources.Requests {
+		r, ok := resources[name]
+		if !ok {
+			r = resource{}
+		}
+		r.request = quantity.String()
+	}
+	for name, res := range resources {
+		switch name {
+		case api.ResourceCPU:
+			name = "resource/cpu"
+		case api.ResourceMemory:
+			name = "resource/memory"
+		default:
+			glog.Warningf("Resource type not supported: %v", name)
+		}
+		value := []byte(fmt.Sprintf(`"name":%q,"value":{"request":%q,"limit":%q}`, name, res.request, res.limit))
```

Comment:
Can we build the struct here instead of unmarshalling JSON? I don't think we have the constructor problem from above.